src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSet.cs

```diff
@@ -197,88 +202,404 @@ private void AddKnownDataContracts(DataContractDictionary? knownDataContracts)
+        [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
+        private void AddReferencedType(Dictionary<XmlQualifiedName, object> referencedTypes, Type type)
         {
-            return Contracts.GetEnumerator();
+            if (IsTypeReferenceable(type))
+            {
+                XmlQualifiedName xmlName;
+                try
+                {
+                    xmlName = GetXmlName(type);
+                }
+                catch (InvalidDataContractException)
+                {
+                    // Type not referenceable if we can't get a xml name.
+                    return;
+                }
+                catch (InvalidOperationException)
+                {
+                    // Type not referenceable if we can't get a xml name.
+                    return;
+                }
+
+                if (referencedTypes.TryGetValue(xmlName, out object? value))
+                {
+                    if (value is Type referencedType)
+                    {
+                        if (referencedType != type)
+                        {
+                            referencedTypes.Remove(xmlName);
+                            List<Type> types = new List<Type>();
+                            types.Add(referencedType);
+                            types.Add(type);
+                            referencedTypes.Add(xmlName, types);
+                        }
+                    }
+                    else
+                    {
+                        List<Type> types = (List<Type>)value;
+                        if (!types.Contains(type))
+                            types.Add(type);
+                    }
+                }
+                else
+                    referencedTypes.Add(xmlName, type);
+            }
+        }
+
+        [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
+        private static bool IsTypeReferenceable(Type type)
+        {
+            try
+            {
+                return (type.IsSerializable ||
+                        type.IsDefined(Globals.TypeOfDataContractAttribute, false) ||
+                        (Globals.TypeOfIXmlSerializable.IsAssignableFrom(type) && !type.IsGenericTypeDefinition) ||
+                        CollectionDataContract.IsCollection(type, out _) ||
+                        ClassDataContract.IsNonAttributedTypeValidForSerialization(type));
+            }
+            catch (Exception)
+            {
+                // An exception can be thrown in the designer when a project has a runtime binding redirection for a referenced assembly or a reference dependent assembly.
+                // Type.IsDefined is known to throw System.IO.FileLoadException.
+                // ClassDataContract.IsNonAttributedTypeValidForSerialization is known to throw System.IO.FileNotFoundException.
+                // We guard against all non-critical exceptions.
+            }
```

Comment:
This code doesn't guard against all non-critical exceptions, it guards against all including critical ones. The original code had `if (Fx.IsFatal(ex)) throw;` which meant critical exceptions would get rethrown. At the very least change the comment, but preferably we shouldn't catch critical exceptions.